d2common/d2math/vector2.go

```diff
@@ -0,0 +1,367 @@
+	deltaSq := delta.Multiply(delta)
+
+	return big.NewFloat(zero).Add(deltaSq.X(), deltaSq.Y())
+}
+
+// Length returns the length of this Vector2
+func (v *Vector2) Length() *big.Float {
+	xsq, ysq := v.LengthSq()
+
+	return xsq.Add(xsq, ysq)
+}
+
+func (v *Vector2) LengthSq() (*big.Float, *big.Float) {
+	clone := v.Clone()
+	x, y := clone.X(), clone.Y()
+
+	return x.Mul(x, x), y.Mul(y, y)
+}
+
+// SetLength sets the length of this Vector2
+func (v *Vector2) SetLength(length *big.Float) d2interface.Vector2 {
+	return v.Normalize().Scale(length)
+}
+
+// Normalize Makes the vector a unit length vector (magnitude of 1) in the same
+// direction.
+func (v *Vector2) Normalize() d2interface.Vector2 {
+	xsq, ysq := v.LengthSq()
+	length := big.NewFloat(zero).Add(xsq, ysq)
+	one := big.NewFloat(1.0)
+
+	if length.Cmp(one) > 0 {
+		length.Quo(one, length.Sqrt(length))
+
+		v.x.Mul(v.x, length)
+		v.y.Mul(v.y, length)
+	}
+
+	return v
+}
+
+// NormalizeRightHand rotate this Vector2 to its perpendicular,
+//in the positive direction.
+func (v *Vector2) NormalizeRightHand() d2interface.Vector2 {
+	x := v.x
+	v.x = v.y.Mul(v.y, big.NewFloat(negative1))
+	v.y = x
+
+	return v
+}
+
+// NormalizeLeftHand rotate this Vector2 to its perpendicular,
+//in the negative1 direction.
+func (v *Vector2) NormalizeLeftHand() d2interface.Vector2 {
+	x := v.x
+	v.x = v.y
+	v.y = x.Mul(x, big.NewFloat(negative1))
+
+	return v
+}
+
+// Calculate the dot product of this Vector2 and the given Vector2
+func (v *Vector2) Dot(src d2interface.Vector2) *big.Float {
+	c := v.Clone()
+	c.X().Mul(c.X(), src.X())
+	c.Y().Mul(c.Y(), src.Y())
+
+	return c.X().Add(c.X(), c.Y())
+}
+
+// Cross Calculate the cross product of this Vector2 and the given Vector2.
+func (v *Vector2) Cross(src d2interface.Vector2) *big.Float {
+	c := v.Clone()
+	c.X().Mul(c.X(), src.X())
+	c.Y().Mul(c.Y(), src.Y())
+
+	return c.X().Sub(c.X(), c.Y())
+}
+
+// Lerp Linearly interpolate between this Vector and the given Vector2.
+func (v *Vector2) Lerp(
+	src d2interface.Vector2,
+	t *big.Float,
+) d2interface.Vector2 {
+	vc, sc := v.Clone(), src.Clone()
+	x, y := vc.X(), vc.Y()
+	v.x.Set(x.Add(x, t.Mul(t, sc.X().Sub(sc.X(), x))))
+	v.y.Set(y.Add(y, t.Mul(t, sc.Y().Sub(sc.Y(), y))))
+
+	return v
+}
+
+// Reset this Vector the zero vector (0, 0).
+func (v *Vector2) Reset() d2interface.Vector2 {
+
+	v.x.SetFloat64(zero)
+	v.y.SetFloat64(zero)
+
+	return v
+}
+
+// Limit the length (or magnitude) of this Vector2
+func (v *Vector2) Limit(max *big.Float) d2interface.Vector2 {
+	length := v.Length()
+
+	if max.Cmp(length) < 0 {
+		v.Scale(length.Quo(max, length))
+	}
+
+	return v
+}
+
+// Reflect this Vector2 off a line defined by a normal.
+func (v *Vector2) Reflect(normal d2interface.Vector2) d2interface.Vector2 {
+	clone := v.Clone()
+	clone.Normalize()
+
+	two := big.NewFloat(2.0) // there's some matrix algebra magic here
+	dot := v.Clone().Dot(normal)
+	normal.Scale(two.Mul(two, dot))
+
+	return v.Subtract(normal)
```

Comment:
Could it be (normal - v) instead of v - normal ?
From Godot::Vector2 `return 2.0 * p_normal * this->dot(p_normal) - *this;`